ArduPlane/soaring.cpp

```diff
@@ -66,156 +66,118 @@ void Plane::update_soaring() {
         // Update thermal estimate and check for switch back to AUTO
         g2.soaring_controller.update_thermalling();  // Update estimate
 
+        // Thermalling is done in a home-relative coordinate system, so we need home to be set.
+        Vector3f position;
+        if (!ahrs.get_relative_position_NED_home(position)) {
+            return;
+        }
+
+        // Check distance to home against MAX_RADIUS.
+        if (g2.soaring_controller.max_radius >= 0 &&
+            powf(position.x,2)+powf(position.y,2) > powf(g2.soaring_controller.max_radius,2) &&
+            previous_mode->mode_number()!=Mode::Number::AUTO) {
+            // Some other loiter status, and outside of maximum soaring radius, and previous mode wasn't AUTO
+            gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Outside SOAR_MAX_RADIUS, RTL");
+            set_mode(mode_rtl, ModeReason::SOARING_DRIFT_EXCEEDED);
+            break;
+        }
+
+        // If previous mode was AUTO and there was a previous NAV command, we can use previous and next wps for drift calculation
+        // with respect to the desired direction of travel. If these vectors are zero, drift will be calculated from thermal start
+        // position only, without taking account of the desired direction of travel.
         Vector2f prev_wp, next_wp;
 
-        // If previous mode was AUTO and there was a previous NAV command, we can use previous and next wps for drift calculation.
-        if (previous_mode->mode_number() == Mode::Number::AUTO) {
+        if (previous_mode == &mode_auto) {
             AP_Mission::Mission_Command current_nav_cmd = mission.get_current_nav_cmd();
             AP_Mission::Mission_Command prev_nav_cmd;
 
             if (!(mission.get_next_nav_cmd(mission.get_prev_nav_cmd_with_wp_index(), prev_nav_cmd) &&
-                   prev_nav_cmd.content.location.get_vector_xy_from_origin_NE(prev_wp) &&
+                prev_nav_cmd.content.location.get_vector_xy_from_origin_NE(prev_wp) &&
                 current_nav_cmd.content.location.get_vector_xy_from_origin_NE(next_wp))) {
-                prev_wp.x = 0.0;
-                prev_wp.y = 0.0;
-                next_wp.x = 0.0;
-                next_wp.y = 0.0;
+                prev_wp.zero();
```

Comment:
If there was some guarantee on `get_vector_xy_from_origin_NE` that its argument would remain unchanged then we wouldn't need to zero here (Vector2 and Vector3 are implicitly zeroed in ArduPilot).  Sadly, there's no such comment in the header ('though the current implementation has the desired property).

Reply:
Left unchanged for now.